rent/src/lib.rs

```diff
@@ -0,0 +1,1489 @@
+					renter_cancellation_nft_id = Some(*nft_id);
+				},
+				_ => (),
+			}
+		}
+		if let Some(rentee_cancellation_fee) = rentee_cancellation_fee {
+			match rentee_cancellation_fee {
+				CancellationFee::NFT(nft_id) => rentee_cancellation_nft_id = Some(*nft_id),
+				_ => (),
+			}
+		}
+		ensure!(nft_id != rent_fee_nft_id, Error::<T>::InvalidFeeNFT);
+		ensure!(nft_id != renter_cancellation_nft_id, Error::<T>::InvalidFeeNFT);
+		ensure!(nft_id != rentee_cancellation_nft_id, Error::<T>::InvalidFeeNFT);
+		if rent_fee_nft_id.is_some() {
+			ensure!(
+				rent_fee_nft_id != nft_id &&
+					rent_fee_nft_id != renter_cancellation_nft_id &&
+					rent_fee_nft_id != rentee_cancellation_nft_id,
+				Error::<T>::InvalidFeeNFT
+			);
+		}
+		if renter_cancellation_nft_id.is_some() {
+			ensure!(
+				renter_cancellation_nft_id != nft_id &&
+					renter_cancellation_nft_id != rent_fee_nft_id &&
+					renter_cancellation_nft_id != rentee_cancellation_nft_id,
+				Error::<T>::InvalidFeeNFT
+			);
+		}
+		if rentee_cancellation_nft_id.is_some() {
+			ensure!(
+				rentee_cancellation_nft_id != nft_id &&
+					rentee_cancellation_nft_id != rent_fee_nft_id &&
+					rentee_cancellation_nft_id != renter_cancellation_nft_id,
+				Error::<T>::InvalidFeeNFT
+			);
+		}
+
+		Ok(())
+	}
+
+	/// Get contract total blocks.
+	pub fn get_contract_total_blocks(duration: &Duration<T::BlockNumber>) -> T::BlockNumber {
+		match duration {
+			Duration::Fixed(value) => *value,
+			Duration::Subscription(value, max_value) =>
+				if let Some(max_value) = *max_value {
+					max_value
+				} else {
+					*value
+				},
+			Duration::Infinite => T::BlockNumber::from(0u32),
+		}
+	}
+
+	/// Get contract end block.
+	pub fn get_contract_end_block(
+		nft_id: NFTId,
+		duration: &Duration<T::BlockNumber>,
+		queues: &mut RentingQueues<T::BlockNumber, T::SimultaneousContractLimit>,
+	) -> Option<T::BlockNumber> {
+		match duration {
+			Duration::Fixed(_) => queues.fixed_queue.get(nft_id),
+			Duration::Subscription(_, _) => queues.subscription_queue.get(nft_id),
+			Duration::Infinite => Some(T::BlockNumber::from(0u32)),
+		}
+	}
+
+	/// Takes the rent fee from the rentee.
+	pub fn take_rent_fee(
+		from: &T::AccountId,
+		to: T::AccountId,
+		rent_fee: &RentFee<BalanceOf<T>>,
+	) -> Result<(), DispatchError> {
+		match rent_fee {
+			RentFee::NFT(nft_id) => {
+				let mut fee_nft =
+					T::NFTExt::get_nft(*nft_id).ok_or(Error::<T>::NFTNotFoundForRentFee)?;
+				ensure!(fee_nft.owner == *from, Error::<T>::NotTheNFTOwnerForRentFee);
+				Self::ensure_nft_available(&fee_nft)?;
+				fee_nft.owner = to;
+				T::NFTExt::set_nft(*nft_id, fee_nft)?;
+			},
+			RentFee::Tokens(amount) => {
+				ensure!(
+					T::Currency::free_balance(&from) >= *amount,
+					Error::<T>::NotEnoughBalanceForRentFee
+				);
+				T::Currency::transfer(&from, &to, *amount, KeepAlive)?;
+			},
+		}
+		Ok(())
+	}
+
+	/// Takes the cancellation fee from renter or rentee.
+	pub fn take_cancellation_fee(
```

Comment:
This can be made more beautiful: 
```rust
	/// Takes the cancellation fee from renter or rentee.
	pub fn take_cancellation_fee(
		from: &T::AccountId,
		cancellation_fee: &Option<CancellationFee<BalanceOf<T>>>,
	) -> Result<(), DispatchError> {
		let cancellation_fee = match cancellation_fee {
			Some(x) => x,
			None => return Ok(()),
		};

		if let Some(amount) = cancellation_fee.get_balance() {
			let free_balance = T::Currency::free_balance(from);
			ensure!(free_balance > amount, Error::<T>::NotEnoughBalanceForCancellationFee);
			T::Currency::transfer(from, &Self::account_id(), amount, KeepAlive)?;
		}

		if let Some(nft_id) = cancellation_fee.get_nft() {
			let mut cancellation_nft =
				T::NFTExt::get_nft(nft_id).ok_or(Error::<T>::NFTNotFoundForCancellationFee)?;
			ensure!(cancellation_nft.owner == *from, Error::<T>::NotTheNFTOwnerForCancellationFee);
			Self::ensure_nft_available(&cancellation_nft)?;
			cancellation_nft.owner = Self::account_id();
			T::NFTExt::set_nft(nft_id, cancellation_nft)?;
		}

		Ok(())
	}
```